### DISCUSSION TITLE
Execute on_start only at start of parent timeline

### DISCUSSION THREAD

**Discussion by dschreij** [jspsych/jsPsych#767] [Q&A] [answered]
I am creating an experiment in analogy of https://groups.google.com/forum/#!topic/jspsych/fCKiH-HRvGM. I want the trials to keep going for a limited amount of time, after which the block is ended. Because I am using `loop_function` to determine if a trial should be repeated. I cannot use `timeline_variables` (as far as I know?) so I have to implement the sampling of the trial variables myself. I have this so far:

```javascript
function init_pool() {
    var conditions = [
        {cue: 'A',          probe: 'X',             trial_type: 'AX'},
        {cue: 'A',          probe: 'X',             trial_type: 'AX'},
        {cue: randomLetter, probe: 'X',             trial_type: 'BX'},
        {cue: 'A',          probe: randomLetter,    trial_type: 'AY'},
        {cue: randomLetter, probe: randomLetter,    trial_type: 'BY'},
    ];
    return jsPsych.randomization.repeat(conditions, 3);
}

var distractors = ['B','C','D','E','F','G','H','I','J','L','M','N','O','P','Q','R',
    'S','T','U','V','W','Y','Z'];

function randomLetter() {
    return distractors[Math.floor(Math.random()*distractors.length)];
}

var trials = {
        timeline: [fixation, first_letter, blank, second_letter, feedback],
        loop_function: function(){
            return block_duration - Date.now() + block_start_time >= 0;
        },
        on_start: function(trial) {
            if (condition_pool.length === 0){
                condition_pool = init_pool();
            }
            var currentTrial = condition_pool.pop();
            // The below obviously doesn't work, but it would be nice ;)
            // I just created global variables for the properties I want to be 'active' for the current trial
            trial.timeline_variables = currentTrial;
        }
}
```
The main problem is that the `on_start` function is propagated to all elements in the timeline, so fixation, first_letter, etc. all execute this on_start function.

Is it also possible to specify a function that only fires at the start of the current element, and not at the start of each of its child components? I tried on_load too, but it has the same behavior.

**Comment by dschreij** [comment, marked as answer]
Oh wait. You can use `timeline_variables`... :facepalm: . I thought it would only repeat the current trial (and its values) and not sample new ones. But I guess in a different context my question still stands. Is it possible to execute a function only at the start of its timeline, and not propagate it to all of its child components?

**Reply by jodeleeuw** [reply]
No, not yet. This is an open issue that we are planning to address with the 7.0 release. #585

**Reply by dschreij** [reply]
Alright, good to know. Thanks!

**Comment by dschreij** [comment]
Alright, I have to revisit this again. I found out that the loop_function is only called after all timeline_variables have been processed, while I hoped it would evaluate after every trial. It seems you cannot use timeline_variables after all in a situation which you want to evaluate after each trial if the block should be stopped or continue depending on the time passed.

**Reply by jodeleeuw** [reply]
This is a helpful use case for us to consider. Maybe adding something like `break_function` to timelines would help.

I think you can use the `on_finish` event of the last trial in the timeline and call `jsPsych.endCurrentTimeline()` ([docs here](https://www.jspsych.org/core_library/jspsych-core/#jspsychendcurrenttimeline)) if you want to break out of that timeline. Caveat: this kind of timeline control hasn't been as thoroughly tested as most features of the library, so it's possible that there are edge cases we haven't caught yet.

**Reply by dschreij** [reply]
Thanks for your suggestion. That's the first approach I took too, but somehow that gave unexpected behavior. It also depends a lot on how deeply you nest your timeline. If you are in timeline X which contains Y, and you want to stop X, then I have found if you call `jsPsych.endCurrentTimeline()` at the moment the experiment is in timeline Y, it only breaks out of that segment, and not out of X. This is well described in the docs, but this makes the usage of `endCurrentTimeline` limited in this sense. I solved it now by just using functions in the 'timed' timeline for all dynamic parts (e.g. which stimulus to display and which response to accept). It's a bit more work, but works perfectly!
